src/libraries/Common/src/System/Net/SocketProtocolSupportPal.Windows.cs

```diff
@@ -15,8 +15,18 @@ internal static class SocketProtocolSupportPal
         public static bool OSSupportsIPv4 { get; } = IsSupported(AddressFamily.InterNetwork);
         public static bool OSSupportsUnixDomainSockets { get; } = IsSupported(AddressFamily.Unix);
 
+        private static bool s_initialized;
+
         private static bool IsSupported(AddressFamily af)
         {
+            // Ensure that WSAStartup has been called once per process.
+            // The System.Net.NameResolution contract is responsible for the initialization.
+            if (!s_initialized)
+            {
+                Dns.GetHostName();
```

Comment:
This will make an extra call `GetHostName` in `System.Net.Sockets`. It can be special cased for `System.Net.Sockets` if we cared about this minor inefficiency.

We are not shutting down winsock during process exit. Is there a good reason why we want to initialize winsock exactly once? I think it would be fine for each .dll to initialize winsock. It would eliminate the unnecessary `GetHostName` call completely.

Reply:
> We are not shutting down winsock during process exit. Is there a good reason why we want to initialize winsock exactly once?

Is WSAStartup safe to be used concurrently?

Reply:
It must be. Otherwise, you cannot have two .dlls in one application that talk to network and do not know about each other.

(I have checked the implementation and the initialization is protected by a big critical section.)

Reply:
Done.